tests/sycl/test_sycl_containers.sycl

```diff
@@ -79,3 +82,80 @@ TEST_F( sycl_containers_test, shared_memory ) {
                  inputvec.at( i ) * constants[ 0 ] + constants[ 1 ] );
    }
 }
+
+/// Test a linear transformation using the host/device memory resources
+TEST_F( sycl_containers_test, device_memory ) {
+
+   // Create the SYCL queue that we'll be using in the test.
+   cl::sycl::queue queue{ vecmem::sycl::device_selector() };
+
+   // The memory resources.
+   vecmem::sycl::host_memory_resource host_resource( &queue );
+   vecmem::sycl::device_memory_resource device_resource( &queue );
+
+   // Helper object for performing memory copies.
+   vecmem::sycl::copy copy( &queue );
+
+   // Create an input and an output vector in host memory.
+   vecmem::vector< int > inputvec( { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 },
+                                   &host_resource );
+   vecmem::vector< int > outputvec( inputvec.size(), &host_resource );
+   EXPECT_EQ( inputvec.size(), outputvec.size() );
+
+   // Allocate a device memory block for the output container.
+   auto outputvechost = vecmem::get_data( outputvec );
+   vecmem::data::vector_buffer< int >
+      outputvecdevice( outputvec.size(), device_resource );
+
+   // Create the array that is used in the linear transformation.
+   vecmem::array< int, 2 > constants( host_resource );
+   constants[ 0 ] = 2;
+   constants[ 1 ] = 3;
+
+   // Explicitly copy the input objects to the device. We need to do it outside
+   // of the @c sycl::queue::submit call, as the local variables of the lambda
+   // given to that function are deleted before the kernel would run.
+   auto const_data = copy.to( vecmem::get_data( constants ),
+                              device_resource );
+   auto input_data = copy.to( vecmem::get_data( inputvec ),
+                              device_resource );
+
+   // Perform a linear transformation using the vecmem vector helper types.
+   queue.submit( [ &const_data, &input_data,
+                   &outputvecdevice ]( cl::sycl::handler& h ) {
+
+      // Run the kernel.
+      h.parallel_for< class LinearTransform2 >(
+         cl::sycl::range< 1 >( input_data.m_size ),
+         [ constants = vecmem::get_data( const_data ),
+           input = vecmem::get_data( input_data ),
+           output = vecmem::get_data( outputvecdevice ) ]( cl::sycl::id< 1 > id ) {
+
+            // Skip invalid indices.
+            if( id >= input.m_size ) {
+               return;
+            }
+
+            // Create the helper containers.
+            const vecmem::const_device_array< int, 2 >
+               constantarray( constants );
+            const vecmem::const_device_vector< int > inputvec( input );
+            vecmem::device_vector< int > outputvec( output );
+
+            // Perform the linear transformation.
+            outputvec.at( id ) = inputvec.at( id ) * constantarray.at( 0 ) +
+                                 constantarray.at( 1 );
+            return;
+         } );
+   } );
+   queue.wait(); // This is needed just for the OpenCL CPU driver...
```

Comment:
It turns out I was incorrect here. :confused: I don't know why this test was/is working correctly for me from inside of Docker with a GPU backend. I suspect that the newer version of oneAPI that I'm using, already supports "asynchronous" copies. But when I use the last oneAPI production release, this `wait()` call must be here for all backends...